### ISSUE TITLE
Printed document does not respect actual font and font size

### ISSUE THREAD

**Issue by fharr** [ClosedXML/ClosedXML#469]
**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature

**Version of ClosedXML**

e.g. 0.88.0

**What is the current behavior?**

I am generating an excel with two worksheets. The first contains some business data, the second one contains the **Terms and Condition** of my company.

When the excel file is done generating, open it and directly go to the print option. Select **print the entire workbook** and scroll down the preview to see the pages of the second worksheet.

You will see that the font and font size of the preview is not the same as the one I setup in my code. If you go back to the file, open the second worksheet and then return to the print options, the preview now  respects the font and font size.

Notice that I encounter the same issue when exporting to PDF or with the actually printed document. But once the second tab has been actually rendered by Excel, everything is back to normal.

Finally, I change my code (the sample below is the first version of the code, with two worksheets) to add the **Terms and Conditions** at the end of the first worksheet. But the problem is the same. If I don't scroll until Excel actually renders the **Terms and Conditions**, the font and font size are not respected during printing or pdf exporting.

I already post a question on [Stack Overflow](https://stackoverflow.com/questions/45982554/generated-excel-with-closedxml-does-not-use-the-specified-font-size-during-print?noredirect=1#comment79128653_45982554) which drives me here.

**What is the expected behavior or new feature?**

I expect the printed or exported document to respect the font and font size I setup in my code without having to force Excel to render if before.

I think it is a problem with merged cells, since I am merging a lots of rows to render my **Terms and Conditions**.

As I said in the Stack Overflow question, I am not sure it is a ClosedXML bug. I am suspecting an Excel bug instead. But may be there is a way to fix it in ClosedXML ?

**Did this work in previous versions of our tool?  Which versions?**

I don't know.

**Code to reproduce problem:**
```c#
class Program
{
	const int nbRowByPage = 67;
	const int nbCharByPage = 3650;

	const string generalTermsFileName = "ConditionsGeneralSales_R5.docx";
	const string baseTemplateFileName = "TemplateOffer.xlsx";
	const string generatedExcelFileName = "Offer.xlsx";

	static void Main(string[] args)
	{
		// Deletes any existing excel file
		var fi = new FileInfo(generatedExcelFileName);
		if (fi.Exists)
		{
			fi.Delete();
		}

		using (var workbookDocument = new XLWorkbook(baseTemplateFileName))
		{
			// Fills first worksheet
			// [..]

			// Gets the general terms from the word and injects them into the second worksheet
			using (WordprocessingDocument wordprocessingDocument = WordprocessingDocument.Open(generalTermsFileName, true))
			{
				// Extracts the paragraphs
				var paragraphs = wordprocessingDocument.MainDocumentPart.Document.Body
					.Descendants<Paragraph>()
					.Where(p => !String.IsNullOrWhiteSpace(p.InnerText))
					.ToList();

				// Gets the second worksheet
				var conditionsWS = workbookDocument.Worksheet(2);

				// Injects the paragraphs in the worksheets
				PrintTerms(conditionsWS, paragraphs);
			}

			// Saves the excel
			workbookDocument.SaveAs(generatedExcelFileName, true);
		}

		// Opens the generated excel with MS Excel
		Process.Start(generatedExcelFileName);
	}

	/// <summary>
	/// Adds the specified paragraph into the specified worksheet according this methodology:
	/// - Splits the paragraphs into two columns 
	/// - Fills the excel columns alternatively in order to looks like a two columns word text
	/// - Manage the break page
	/// </summary>
	/// <param name="ws"></param>
	/// <param name="generalTerms"></param>
	static void PrintTerms(IXLWorksheet ws, List<Paragraph> generalTerms)
	{
		var currentParagraphIndex = 0;
		var index = 1;

		while (currentParagraphIndex < generalTerms.Count)
		{
			var nbChar = 0;
			var currentParagraphs = generalTerms.Skip(currentParagraphIndex).TakeWhile(p => (nbChar += p.InnerText.Length) < nbCharByPage).ToList();

			var row = ((int)Math.Ceiling(index / 2d) - 1) * nbRowByPage + 1;
			var column = index % 2 == 0 ? 3 : 1;

			var currentCell = ws.Cell(row, column);
			currentCell.Style.Alignment.SetHorizontal(XLAlignmentHorizontalValues.Justify);
			currentCell.Style.Alignment.SetVertical(XLAlignmentVerticalValues.Top);
			currentCell.Style.Font.SetFontName("Arial");
			currentCell.Style.Font.SetFontSize(8);

			foreach (var paragraph in currentParagraphs)
			{
				foreach (var run in paragraph.Descendants<Run>())
				{
					var text = currentCell.RichText.AddText(run.InnerText);

					if (run.RunProperties.Bold != null)
					{
						text.SetBold();
					}

					if (run.RunProperties.Italic != null)
					{
						text.SetItalic();
					}

					if (run.RunProperties.Underline != null)
					{
						text.SetUnderline();
					}
				}

				// Break line
				currentCell.RichText.AddNewLine();
				currentCell.RichText.AddNewLine();
			}

			ws.Range(row, column, row + nbRowByPage - 1, column).Merge();

			index++;
			currentParagraphIndex += currentParagraphs.Count();
		}
	}
}
```
Here are two files needed to run the code. Please, check copy to bin folder before running.

[ConditionsGeneralSales_R5.docx](https://github.com/ClosedXML/ClosedXML/files/1284041/ConditionsGeneralSales_R5.docx)
[TemplateOffer.xlsx](https://github.com/ClosedXML/ClosedXML/files/1284040/TemplateOffer.xlsx)

**Comment by vbjay** [comment]
You didn't link the dependency for the wordprocessingdocument class.

**Comment by fharr** [comment]
WordprocessingDocument is part of OpenXML (DocumentFormat.OpenXml). If you install ClosedXml through nuget, it will be installed in the same time. Otherwise, you can install it with [nuget](https://www.nuget.org/packages/DocumentFormat.OpenXml/). 

You just have to add these usings:

```c#
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;
``

**Comment by vbjay** [comment]
I see.

**Comment by igitur** [comment]
Can you try reducing the code sample to the smallest, complete verifiable example to reproduce the problem? For example, get rid of the Word stuff. You can hardcode of pieces of text in your code.

**Comment by fharr** [comment]
I tried to shorten the code. I hope it will be good :D.

**Comment by fharr** [comment]
I also changed the OfferTemplate. Please use the new version.

**Comment by igitur** [comment]
I ran your code and can't replicate the issue. Have you tried on a different machine with the latest version of Excel? And to be sure, try v0.89.0-beta1 of ClosedXML, but I don't think that's relevant.

**Comment by nefcanto** [comment]
I'm having the same issue. I use complex fonts for Persian language. I set font using `cell.Style.Font.FontName = "B Yagut"`, I also tried `SetFontName` method, and I also tried with default styles. None would work for me. Even after opening the Excel file, and rendering the content, font is not applied, until I manually change font to something and then change it back.

Here's what I did. I noticed that when I open the file with Excel, and just save it without any modification the result XML documents will differ from what ClosedXML has generated. In other words do as described below:

- Create an Excel file using ClosedXML
- Using 7zip extract the contents into folder 1
- Open the file, just save it another time without altering anything
- Again using 7zip extract contents into folder 2
- Now compare contents using a folder comparison software (I'm using Beyond Compare)

You see files are different.

I have no idea what to do to apply fonts on Excel file. Please help.

**Comment by igitur** [comment]
Yes, the files are very different from Excel. It was never the goal to produce a binary equivalent.

I can't reproduce it the problem, so there is nothing I can do yet.

What version of Excel are you using? Is it up to date with patches? Have you tried it on a different machine?

**Comment by fharr** [comment]
The version of Office used within my company is : MS Office Standard 2013 (15.0.4569.1506). My application targets the version 4.5.2 of the .NET Framework and I can't reproduce the issue on Windows 10 computers. Only on computers running Windows 7.

Unless nefcanto, I am using a very common font: Arial.

**Comment by igitur** [comment]
I'm also on Office 2013 / Windows 7 - and can't reproduce. @fharr , can you please produce the file with ClosedXML, but don't open it with Excel. Attach it here. 

Additional request: if you can do a short desktop video recording of you reproducing the process. Maybe I'm missing a step or clicking somewhere where I shouldn't.

**Comment by fharr** [comment]
[Here](https://github.com/ClosedXML/ClosedXML/files/1291885/Offer.xlsx) is the generated file. I also linked a powerpoint to guide you through the [process](https://github.com/ClosedXML/ClosedXML/files/1291933/Process.pptx).

**Comment by igitur** [comment]
Thanks @fharr . I followed the steps exactly and couldn't reproduce. Each time I end up with your "2nd version" of the print preview.

My Office versions:
![image](https://user-images.githubusercontent.com/145854/30267135-dee51074-96e1-11e7-8c70-74038ce70ff9.png)

Not sure what else to do.

**Comment by fharr** [comment]
For the same version (standard 2013), it works on W10 and not on W7. With your professional version on W7, it also works. I think it confirms it is a bug in the MS Office suite. Looks like I am fucked :D. Thanks for your time.

**Comment by igitur** [comment]
How about you make the 2nd sheet the active sheet, so that it is displayed by default when you open the workbook?

**Comment by fharr** [comment]
We tried that too, but it was not satisfactory because if the user switched to the first worksheet, saved and reopened the file later, the issue is still here.

**Comment by igitur** [comment]
So you can reproduce the issue even AFTER saving it in Excel?

**Comment by fharr** [comment]
Yes. When I open/close the file, Excel does not ask even for saving the changes.

**Comment by igitur** [comment]
But if you don't change anything in Excel, but you explicitly save the file and reopen it, can you reproduce the issue?

**Comment by fharr** [comment]
Even if I change something in the file, save it, reopen it. As long as, during the current "excel session", I don't actually render the merged cells, I have the issue. After rendering the cells, if I close the file, Excel does not ask for saving the changes.

I said merged cell because even if I put the General Terms in the first worksheet, as long as they are not rendered at the opening (basically because there is too many generated lines before), I get the issue. That's why I think the problem is about the merged cells.

**Comment by igitur** [comment]
If you can produce 2 files (both AFTER an explicit save in Excel), but 1 reproducing the problem and the other 1 being a "fixed" version, then I can do an internal diff and maybe spot the cause. If you have to change cell contents to produce the fixed version, that's Ok, just try to keep the changes to a minimum, else the diff is too big.

**Comment by fharr** [comment]
[This](https://github.com/ClosedXML/ClosedXML/files/1292444/Actual.xlsx) file is the original file but saved once in excel. When I open it and try to print, I have the issue.

[This](https://github.com/ClosedXML/ClosedXML/files/1292446/Expected.xlsx) file is a copy of the previous one where I manually set the police of the first merged range in the second worksheet to 8 (was already to 8, so just a useless change in theory). Notice that only that first merged range (it represent one printed page) is well displayed in the print preview. The second page (corresponding to the second merged range) is still wrong "formatted".

**Comment by igitur** [comment]
[file1.xlsx](https://github.com/ClosedXML/ClosedXML/files/1292480/file1.xlsx)
[file2.xlsx](https://github.com/ClosedXML/ClosedXML/files/1292481/file2.xlsx)
[file3.xlsx](https://github.com/ClosedXML/ClosedXML/files/1292482/file3.xlsx)
[file4.xlsx](https://github.com/ClosedXML/ClosedXML/files/1292483/file4.xlsx)
[file5.xlsx](https://github.com/ClosedXML/ClosedXML/files/1292484/file5.xlsx)
[file6.xlsx](https://github.com/ClosedXML/ClosedXML/files/1292485/file6.xlsx)
[file7.xlsx](https://github.com/ClosedXML/ClosedXML/files/1292486/file7.xlsx)
[file8.xlsx](https://github.com/ClosedXML/ClosedXML/files/1292487/file8.xlsx)

Open these files in numerical order and tell me which ones it the FIRST to reproduce the problem.

**Comment by fharr** [comment]
file1to5: 1st page (range of merged cells) of the 2nd worksheet does not reproduce the issue, 2nd page does.
file6to8: all the pages of the second worksheet have the issue

**Comment by igitur** [comment]
@fharr Ok, we're making progress - you just have to be patient. Are you perhaps on Google Hangouts? If so, pop me a message. My username is same on Gmail as here on Github.

**Comment by nefcanto** [comment]
Guys, sorry to be late. Here is a simple program, and the result in gif:

```
    class Program
    {
        static void Main(string[] args)
        {
            using (var workbook = new ClosedXML.Excel.XLWorkbook())
            {
                var sheet = workbook.Worksheets.Add("Test");

                var a1 = sheet.Cells("A1");
                a1.Value = "سلام";
                a1.Style.Font.FontName = "Tahoma";
                a1.Style.Font.FontSize = 36;

                var a2 = sheet.Cells("A2");
                a2.Value = "سلام";
                a2.Style.Font.FontName = "B Yagut";
                a2.Style.Font.FontSize = 36;

                var a3 = sheet.Cells("A3");
                a3.Value = "سلام";
                a3.Style.Font.FontSize = 36;

                sheet.Columns("A").AdjustToContents();
                workbook.SaveAs(@"C:\Temp\ClosedXmlFontTest.xlsx");
            }
        }
    }
```

Here's the result file, it's NOT opened yet:
[ClosedXmlFontTest.xlsx](https://github.com/ClosedXML/ClosedXML/files/1293213/ClosedXmlFontTest.xlsx)

And here's the gif, showing it on my desktop (opened another copy, not the one attached here):
![closedxmlfonttest](https://user-images.githubusercontent.com/12782347/30284653-c4167086-9730-11e7-9d0d-704b31908f9c.gif)

Excel version: 2016 MSO (16.0.4266.1001) 64-bit
Windows version: 10 Enterprise 64-bit

**Comment by igitur** [comment]
@nefcanto The issue you're describing is unrelated to @fharr 's issue. You have to set the font family numbering correctly:

```c#
private static void Main(string[] args)
{
    using (var workbook = new ClosedXML.Excel.XLWorkbook())
    {
        var sheet = workbook.Worksheets.Add("Test");

        var a1 = sheet.Cells("A1");
        a1.Value = "سلام";
        a1.Style.Font.FontName = "Tahoma";
        a1.Style.Font.FontSize = 36;

        var a2 = sheet.Cells("A2");
        a2.Value = "سلام";
        a2.Style.Font.FontName = "B Yagut";
        a2.Style.Font.FontSize = 36;
        a2.Style.Font.FontFamilyNumbering = XLFontFamilyNumberingValues.NotApplicable;

        var a3 = sheet.Cells("A3");
        a3.Value = "سلام";
        a3.Style.Font.FontSize = 36;

        sheet.Columns("A").AdjustToContents();
        workbook.SaveAs(@"ClosedXmlFontTest.xlsx");
    }

    // Opens the generated excel with MS Excel
    Console.WriteLine("Press any key to continue");
    Console.ReadKey();
}
```

**Comment by nefcanto** [comment]
No change. And by the way, AMAIU from that enumeration, it's for determining whether font should affect numbers or not. In my case, it doesn't affect the entire value including alphabetical characters at all. No change with that line of code happened for me.

**Comment by igitur** [comment]
@nefcanto - Please create a new issue. It's unrelated to this one.

**Comment by igitur** [comment]
@fharr What version of `DocumentFormat.OpenXml` are you using? Please upgrade to `2.7.2` if you're not already on it.

**Comment by fharr** [comment]
Already on `2.7.2`.

**Comment by igitur** [comment]
Styles have had a major rework. Please try again with the latest versions of ClosedXML (0.93.0-beta2) and OpenXML (2.8.2).

**Comment by fharr** [comment]
With this new version, a part of my code producing a merge of cells does not work anymore. Here is the code:

```
var cell = ws.Row(startingRow).Cell(1);
// code puting a long string into the cell, using cell.RichText.AddText(text) a lots of times
ws.Range(startingRow, 1, startingRow + nbRowByPage * nbPage - 1, nbTotalColumn).Merge();
```

Here are the values of thoses variables in my test:

startingRow = 18
nbRowByPage = 68
nbPage = 3
nbTotalColumn = 9

So I want to merge the rows 18 to 204, on the 9 first columns. It actually does nothing, all the text is added into the first cell with no merge.

Is it a regression of the API or a change in the way to use it? 

At some other place in my code, the merge works, but I am not using the same method: `row.Row(headerReferenceColumn, headerDescriptionColumn - 1).Merge();`

I also noticed that the IXLWorsheet's `IXLWorksheet CopyTo(XLWorkbook workbook, string newSheetName);` method does not perfectly copies the content to the new workbook. The height of two merged range (several columns on one row) is not kept in the copy.

**Comment by igitur** [comment]
With reference to the original issue here, about the printed style vs display style, has this been solved in the latest beta version?

**Comment by fharr** [comment]
As I said, the wrongly printed part is within the merged cells (which are no longer merged), so it is hard to tell whether the issue is solved for now.

However, if the merge does not cause this issue, it looks like solved (no change before/after actual preview within the workbook).

**Comment by igitur** [comment]
Ok, let's wait for #829 to be fixed first, then we get back to this one.
